crates/sui-types/src/base_types.rs

```diff
@@ -94,20 +105,178 @@ pub fn random_object_ref() -> ObjectRef {
+    pub fn is_coin(&self) -> bool {
+        match self {
+            MoveType::GasCoin | MoveType::Coin { .. } => true,
+            MoveType::StakedSui | MoveType::Other(_) => false,
+        }
+    }
+
+    /// Return true if `self` is 0x2::coin::Coin<0x2::sui::SUI>
+    pub fn is_gas_coin(&self) -> bool {
+        match self {
+            MoveType::GasCoin => true,
+            MoveType::StakedSui | MoveType::Coin { .. } | MoveType::Other(_) => false,
+        }
+    }
+
+    pub fn is_locked_coin(&self) -> bool {
+        match self {
+            MoveType::GasCoin | MoveType::StakedSui | MoveType::Coin { .. } => false,
+            MoveType::Other(s) => LockedCoin::is_locked_coin(s),
+        }
+    }
+
+    pub fn is_coin_metadata(&self) -> bool {
+        match self {
+            MoveType::GasCoin | MoveType::StakedSui | MoveType::Coin { .. } => false,
+            MoveType::Other(s) => CoinMetadata::is_coin_metadata(s),
+        }
+    }
+
+    pub fn is_dynamic_field(&self) -> bool {
+        match self {
+            MoveType::GasCoin | MoveType::StakedSui | MoveType::Coin { .. } => false,
+            MoveType::Other(s) => DynamicFieldInfo::is_dynamic_field(s),
+        }
+    }
+
+    pub fn try_extract_field_name(&self, type_: &DynamicFieldType) -> SuiResult<TypeTag> {
+        match self {
+            MoveType::GasCoin | MoveType::StakedSui | MoveType::Coin { .. } => {
+                Err(SuiError::ObjectDeserializationError {
+                    error: format!("Error extracting dynamic object name from Coin object"),
+                })
+            }
+            MoveType::Other(s) => DynamicFieldInfo::try_extract_field_name(s, type_),
+        }
+    }
+
+    pub fn is(&self, s: &StructTag) -> bool {
+        match self {
+            MoveType::GasCoin => GasCoin::is_gas_coin(s),
+            MoveType::StakedSui => StakedSui::is_staked_sui(s),
+            MoveType::Coin { inner } => {
+                Coin::is_coin(s) && s.type_params.len() == 1 && inner == &s.type_params[0]
+            }
+            MoveType::Other(o) => s == o,
+        }
+    }
+}
+
+impl From<StructTag> for MoveType {
+    fn from(mut s: StructTag) -> Self {
+        if GasCoin::is_gas_coin(&s) {
+            MoveType::GasCoin
+        } else if Coin::is_coin(&s) {
+            // unwrap safe because a coin has exactly one type parameter
+            MoveType::Coin {
+                inner: s.type_params.pop().unwrap(),
+            }
+        } else {
+            MoveType::Other(s)
+        }
+    }
+}
+
+impl Into<StructTag> for MoveType {
+    fn into(self) -> StructTag {
+        match self {
+            MoveType::GasCoin => GasCoin::type_(),
+            MoveType::StakedSui => StakedSui::type_(),
+            MoveType::Coin { inner } => Coin::type_(inner),
+            MoveType::Other(s) => s,
+        }
+    }
+}
+
+impl Into<TypeTag> for MoveType {
+    fn into(self) -> TypeTag {
+        TypeTag::Struct(Box::new(self.into()))
+    }
+}
+
```

Comment:
Might want to add `impl PartialEq<StructTag> for MoveType` and `impl PartialEq<TypeTag> for MoveType`. That will prevent having to allocate just for `==` and `!=`

Reply:
Ah, I see you have `is` above. Could keep that and also just add the trait, should help reduce some of the lines changed too